Sandcastle - Nested gallery structure

jjspace opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

Major changes/focus for this PR:

- **All sandcastle JS code is now treated as a single JS `module`** that should import `Cesium` and `Sandcastle` directly
    - if the `import` statements are not in the sandcastle code they will be added before loaded into the `bucket`. AFAICT this should account for all existing sandcastles that relied on the global objects
    - This also means these objects are not in global `window` scope anymore
    - Most of the "boilerplate" that previously existed was to set up the `window.startup` function specifically to account for modules loading out of order. Switching to `type=module` eliminates that
    - The types importing for Monaco was updated so that it should support intellisense for `Cesium` and `Sandcastle` regardless of the imports
- **New structure for the gallery** which splits the code into dedicated files and relies on a new "metadata sidecar yaml" file for info about a sandcastle
    - `index.html` is the html/css code for a sandcastle, `main.js` is the JS code for a sandcastle, `sandcastle.yaml` holds the metadata and is what dictates that a subfolder is even a gallery example in the first place
    - Note that `index.html` does _not_ contain the full page structure with `head` and `body` tags. This means you **can no longer load gallery items directly** with a path like `localhost:8080/Apps/Sandcastle/gallery/3D Models.html`.
        - <sup>we feel this was not utilized much and restricted the format too much, maybe we can revisit a way to load them outside of the full SC app in the future</sup>
    - a new script called `buildGallery.js` has been created to parse through the new gallery structure and generate a top level `list.json` file. This also does a little basic validation. Where this script lives and exactly what it does is still a little up in the air

TODO:

- [ ] Clean up the gallery build script if we're satisfied with the organization structure
    - [ ] create a "conversion script" to transform all our existing gallery examples
- [ ] Fix loading from a url id
    - Currently the app does not work in a very "react style" and uses a lot of imperative code which is massively conflicting with load order. This is my next focus as a dedicated restructuring pass. after _that_ I think loading will be a lot easier to implement
- [ ] Decide on gallery location
    - I _think_ the gallery could live outside of the sandcastle project if we wanted. Given [recent comments](https://community.cesium.com/t/upgrading-sandcastle-we-need-your-input/39715/15) around the desire to be able to point at any gallery directory I built the logic to hopefully be able to point to any directory from the client. More thought still required but it's in my mind.

<!-- Describe your changes in detail -->

<!-- Consider: Why is this change required? What problem does it solve? -->

<!-- Include screenshots if appropriate -->

## Issue number and link

Part of #12566 

<!-- If it fixes an open issue, link to the issue here -->

<!-- Consider: If suggesting a new feature or change, discuss it in an issue first. -->

## Testing plan

- run `node buildGallery.js` to generate the list file
    - [ ] need to figure out how to incorporate this into the build nicer
- Load sandcastle both with `npm run dev` and from the normal server `npm start` after `npm run build`
- Make sure the gallery still loads and clicking them loads the examples
    - [ ] may want to consider better where these files live for access from both locations and when deployed
- Make sure types in monaco still show up, with and without the import statements
- Copy in the sandcastle code from another sandcastle and make sure it still works

<!-- Describe in detail how you tested your changes. If this fixes a bug, list the steps to reproduce the original issue. -->

# Author checklist

- [ ] I have submitted a Contributor License Agreement
- [ ] I have added my name to `CONTRIBUTORS.md`
- [ ] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [ ] I have updated the inline documentation, and included code examples where relevant
- [ ] I have performed a self-review of my code

jjspace commented:
@ggetz I tried to split the commits a little between the new gallery files and actual code changes if that makes it easier to review

jjspace commented on packages/sandcastle/buildGallery.js:
```diff
@@ -0,0 +1,111 @@
+
+const galleryDirectory = "./public/gallery";
+
+const yamlFiles = globbySync([
+  `${galleryDirectory}/*/sandcastle.yaml`,
+  `!${galleryDirectory}/list.json`,
+]);
```
Going with the `globby` approach here instead of just listing the directories lets us "anchor" the existence of a Gallery example around a `sandcastle.yaml` file.

jjspace commented on packages/sandcastle/public/gallery/billboards/sandcastle.yaml:
```diff
@@ -0,0 +1,10 @@
+id: billboards
+legacy-id: Billboards.html
+title: Billboards
+thumbnail: ./Billboards.jpg
```
Adding the `thumbnail` property here lets it be named anything. We had discussed maybe not including this and consistently requiring a `thumbnail.png` or something the same as I am for `index.html` and `main.js`. Not sure which approach is better here.
The value is optional regardless if we check for a file or it's defined here.

jjspace commented on packages/sandcastle/src/App.tsx:
```diff
@@ -639,9 +677,19 @@ Sandcastle.addToolbarMenu(${variableName});`,
-  function loadDemo(demo: GalleryDemo) {
-    // do stuff
-    setCode(demo.js ?? defaultJsCode, demo.html ?? defaultHtmlCode);
+  const GALLERY_BASE = "/gallery";
```
I _think_ that doing something like this would allow us to point it at literally any path as long as it conforms to the structure we'd expect. Namely that there's a `list.json` at the root and `[slug]/(index.html|main.js)` files. This could help us swap out where it points for different local/deployed builds or even potentially a different server entirely if someone wants to host their own gallery?

jjspace commented:
@ggetz I know there's still some scattered commented code and console logs. I'll do a cleanup pass on those before this gets merged but wanted to just validate the approach first.

ggetz commented on packages/sandcastle/buildGallery.js:
```diff
@@ -0,0 +1,111 @@
+
+  // Validate metadata
+
+  const slug = basename(dirname(filePath));
```
Maybe use [`slugify`](https://www.npmjs.com/package/slugify) to ensure we're covering our bases, and using that as the `id` directly. That way, there's less room for user error.

Also using the slugified version of the path would also be preferable to make sure we don't get some of the ugly URLs we do now, like `3D%20Tiles%20Clipping%20Planes.html`.

jjspace replied:
@ggetz I like the idea of using `slugify` as part of the conversion script from the existing gallery to the new one. However I don't think we need to use it here. I'd rather the files be as static as possible which means they should match the slug format in the first place. If it doesn't the developer should fix it and this build script will be noisy about it. 

The FE needs to know where to access the files and images. The id is used in that route `[base page]/gallery/[id]`. If we were to generate the id as a `slugify` of the directory name it may not match.

ggetz commented on packages/sandcastle/buildGallery.js:
```diff
@@ -0,0 +1,111 @@
+    title: title,
+    thumbnail: thumbnail,
+    description: description,
+    isNew: false,
```
What the plan for `isNew`?

jjspace replied:
I think I was thinking this could highlight new sandcastles per release. I thought we did that already but I don't think we actually do and I'm not sure where I got that idea. Either way the more I think about it the more I'm not sure it's useful, at least without _fully_ automating it. Probably better to just stick to a Showcase tag and make sure to keep that updated

jjspace commented on packages/sandcastle/templates/Sandcastle.ts:
```diff
@@ -45,7 +45,7 @@ const Sandcastle = {
       let pos = stack.indexOf(needle);
       if (pos < 0) {
-        needle = " (<anonymous>:"; // Chrome
+        needle = "<anonymous>:"; // Chrome
```
I didn't dig into the reason too far but something made the stack output change and it wasn't detecting the line anymore. I _think_ it's because we switched to modules. Either way dropping the `(` will work with both forms.

This whole process needs a validation pass across browsers. I was planning to do it whenever I re-implement the console in the app.

ggetz commented on packages/sandcastle/src/main.tsx:
```diff
@@ -3,6 +3,13 @@ import { createRoot } from "react-dom/client";
+if (host.includes("localhost")) {
+  document.title = `${host.replace("localhost:", "")} ${document.title}`;
+} else if (host.includes("ci")) {
+  document.title = `CI ${document.title}`;
```
I think even in cases where something is deployed to branch, there's still cases where the Sandcastle is used for demo purposed. For instance, the recent Gaussian Splats work is being shown in example Sandcastles from that branch.

jjspace commented on packages/sandcastle/templates/bucket.html:
```diff
@@ -21,24 +29,23 @@
     </script>
-    <!-- <script type="module" src="./Sandcastle.js"></script> -->
     <script src="__CESIUM_BASE_URL__/Cesium.js" vite-ignore></script>
     <script type="module">
```
It still needs to be set somewhere and in a file that's processed by Vite to get the correct path based on the env it's building for. I think this is still the best way to do that. Let me know if you have a better idea.

jjspace commented:
@ggetz I've updated this PR to address most of your comments and responded to others. I think there's still a bit of cleanup I have yet to do but hopefully this is closer.

The gallery has been solidified a bit and the route is now changed depending on the environment so [the CI build](https://ci-builds.cesium.com/cesium/nested-gallery/Apps/Sandcastle2/index.html) now has a working gallery again. It also no longer gets copied and duplicated into the local build, instead it points to the `packages/sandcastle/gallery` directory so local edits will always be reflected immediately.

jjspace commented on packages/sandcastle/src/App.tsx:
```diff
@@ -326,20 +327,59 @@ function App() {
+    } else if (searchParams.has("id")) {
+      const galleryId = searchParams.get("id");
+      console.log("load id", galleryId);
+      // TODO: there's probably a race condition here where the list might not be loaded yet
+      // I need to switch this to the more React declarative style
```
This part is already addressed in #12639

javagl commented:
The `buildSandcastles.js` causes some hiccups on Windows.

- `import.meta.url` is `file:///C:/cesium/packages/sandcastle/scripts/buildGallery.js`
- `process.argv[1]` is `C:\cesium\packages\sandcastle\scripts\buildGallery.js`

causing it to not even enter `buildGalleryList`. When "fixing" that pragmatically, https://github.com/sindresorhus/globby/issues/155 kicks in. Investigating some options...

javagl commented:
Beyond the comment above, I'm not in the position to do a profound _technical_ review (regarding vite, tsx, or CI details). 

Details that may already be on the radar:

- It's necessary to run `npm run buld-app` in the `package/sandcastle` directory (that may be evident, or OBE when this is merged, and maybe the documentation is or will be updated, but I just ran the usual `npm run start` and was surprised that I could open the Sandcastle, but not the Sandcastle2)
- It's necessary to explicitly run `npm run build-ts` (because the code completion needs the TS part, apparently)

---

Some comments that may not belong here, but of which I don't know where they belong (maybe https://github.com/CesiumGS/cesium/pull/12639 ?)

- The new "Add... " buttons are _really_ useful. 
  - One thing that I'm a big fan of, and that I've always been missing in the old sandcastle, are _sliders_. Dragging and seeing the change just makes me feel so connected to my little computing machine 🤗 . It would be great if that could go on the medium-term radar. like `Sandcastle.addSlider("Slide!", min, max, value, function(currentValue) {...})`. 
  - The "Add..." buttons are inserting at the bottom of the code. Could they insert at the current cursor position? 
- I tried the basic parts of code completion, and I'm a big fan of that as well
  - Pressing key sequences like "A [CTRL+Space] . S [ DOWN, ENTER ] ( n [CTRL+Space] [ENTER] )" to write a line like `applicationWindow.showInputDialog(nameProvider)` just feels _so_ more efficient than "real typing"...
  - The help only appears when one of the dropdown options is selected. This is the same as for VSCode. I'm probably just [spoiled](https://github.com/user-attachments/assets/59cb394d-94ac-43e2-8c81-14c9de350a56)...
- The fact that the HTML panel is always visible below the code panel might be hard to get used to...
- The window elements are all fixed size. I assume/hope that they will (all) become such "draggable splitters" like in the old sandcastle

---

Some comments that _may_ belong here:

I have a bunch of Sandcastles from Forum support. They are just a bunch of `.js` files in a directory. (And sometimes HTML, when I needed a slider 😆 ). They follow the name pattern
`Sandcastle Forum <threadId> <short summary>.js`
And I wrote a short script that just squeezes them into the required gallery structure: Create directories called `short-summary`, copy the `.js` file there as `main.js`, add some dummy thumbnail and `index.html`, and a `sandcastle.yaml`. And the running dev server immediately picked them up:

![Cesium SandcastleV2 0001](https://github.com/user-attachments/assets/af90f4bf-5391-4fed-ad16-3dbbbe5c04c3)

Yeah, the thumbnail layout is not perfect (and there is no horizontal scrollbar), but the technical structure seems to work nicely. 

It's not yet clear how exactly the old sandcastles will be "migrated" into the new structure. But given that the _target_ structure is relatively simply (with clearly separated files, and each directory being "one sandcaslte"), this shouldn't be too hard.

ggetz commented on packages/sandcastle/templates/tsconfig.lib.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "include": ["./Sandcastle.ts"],
+  "compilerOptions": {
+    "tsBuildInfoFile": "../node_modules/.tmp/tsconfig.lib.tsbuildinfo",
```
So why do we need to set [`tsBuildInfoFile`](https://www.typescriptlang.org/tsconfig/#tsBuildInfoFile)? And why is it using a path in `node_modules`?

jjspace replied:
Without setting it it defaults to output this file in the same directory. I set it to inside the node modules just matching the pattern in the main `tsconfig` that was generated from the vite setup. the `node_modules` is already `gitignored` so it seems like a reasonable place to put some of these extra build related temp files.

ggetz replied:
It's generally discouraged to write or modify any content to the `node_modules` folder, as it is inherently transitory. Please use another location and add it to the `.gitignore` as necessary. 

What's the issue that arises when that file is written to the current directory? By default, this is written to next to the "out" location, so it may make sense to adjust the output location instead.

jjspace replied:
> It's generally discouraged to write or modify any content to the node_modules folder, as it is inherently transitory.

Generally I would agree with you. However this is following the example set by the default `tsconfig` set up with the [vite template](https://vite.dev/guide/). As I understand it the `tsbuildinfo` file is also pretty temporary and only there to help speed up subsequent local builds.

> What's the issue that arises when that file is written to the current directory?

Just extra noise in the file tree and needing to add it specifically to the gitignore.

jjspace replied:
> But this is not built using vite, correct? It's built with tsc.

It _is_ built with vite when running the local dev server to allow for quicker updates when actively developing sandcastle directly.

> What vite does internally isn't so much my concern– But making sure our project follows good practices and is maintainable is. I'm concerned this approach is obscuring details by doing things in a non-conventional way, making things more difficult to understand and maintain for others in the future.

The point I was trying to make is that the method I was using (tmp dir in node_modules) is what the vite scaffold project does by default. I extrapolated this to mean that that _is_ the conventional or expected way to handle these files and is a "good practice". At least the expected way within vite projects. 

Regardless, I've moved them all next to their configs by just removing the config path. It's easy to change later if we want

ggetz commented on packages/sandcastle/src/Helpers.ts:
```diff
@@ -1,20 +1,46 @@
+
+  let imports = "";
+
+  if (!/import.*Cesium.*from.*;?$/m.test(code)) {
```
But to some extend we _do_ care what's between those tokens. Right now, this regex has the potential match a lot of unexpected things:

`importCesiumfrom`
`import CesiumSomething from SomethingCompletelyDifferent;`
`import something from Cesium; import somethingElse from anotherThing;`
`import(Cesium); myVariablewithfrom`

which are all really different from what we're looking for, right?

In general, [using the greedy `.*` is generally recommended against](https://unix.stackexchange.com/questions/441927/how-or-why-using-is-better-than).

ggetz commented on packages/sandcastle/scripts/buildGallery.js:
```diff
@@ -0,0 +1,139 @@
+import { exit } from "process";
+import { fileURLToPath, pathToFileURL } from "node:url";
+
+// TODO: probably need to find a way to integrate this into the vite process to make sure it's
+// built during build
+// May also want to move this out of the sandcastle project so we don't have to duplicate
+// files for the local build, need to think about deploying too...
```
So can we remove this TODO now that it's brought into the root `build` file?

ggetz commented on packages/sandcastle/buildGallery.js:
```diff
@@ -0,0 +1,111 @@
+const galleryDirectory = "./public/gallery";
+
+const yamlFiles = globbySync([
+  `${galleryDirectory}/*/sandcastle.yaml`,
```
@jjspace What are your thoughts on this comment?

ggetz commented:
I'm still attempting to test this out, by npm appears to have other plans...

<img width="890" alt="image" src="https://github.com/user-attachments/assets/ae932126-bf76-4483-80dc-142cd4545ea1" />

ggetz commented:
Otherwise, I tested this out by running the following commands and things appear to be running smoothly.

```
npm run build
npm run build-sandcastle
npm start
```

jjspace commented on packages/sandcastle/package.json:
```diff
@@ -23,8 +24,10 @@
     "@types/react-dom": "^19.0.4",
     "@vitejs/plugin-react": "^4.3.4",
     "globals": "^15.15.0",
+    "globby": "^14.1.0",
```
> The built app is fetching `list.json` at runtime, correct?

The built app is fetching `gallery-list.json` as a _static file_ at runtime. It's not generated "on the fly" using the functions in `buildGallery`.

ggetz replied:
I still disagree since we're generating `gallery-list` "on the fly" in our root dev server— But I admit its pedantic and won't bikeshed it.

jjspace commented:
@ggetz Updated, please take another look. I think it should be good to go

ggetz commented:
Thanks @jjspace, especially for your patience! 😬
